src/mono/wasm/runtime/http.ts

```diff
@@ -65,79 +65,36 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
     }
 }
 
-export function http_wasm_readable_stream_controller_enqueue(pull_state: PullStateExtension, bufferPtr: VoidPtr, bufferLength: number): void {
-    const controller = pull_state.__controller;
-    const pull_promise_control = pull_state.__pull_promise_control;
-    mono_assert(controller, "expected controller");
-    mono_assert(pull_promise_control, "expected pull_promise_control");
-    try {
-        if (bufferLength === 0) {
-            controller.close();
-            pull_state.dispose();
-            pull_state.__pull_promise_control = null;
-            pull_state.__controller = null;
-        } else {
-            // the bufferPtr is pinned by the caller
-            const view = new Span(bufferPtr, bufferLength, MemoryViewType.Byte);
-            // because https://github.com/WebAssembly/design/issues/1162 we need to copy the buffer
-            // also it doesn't make much sense to use byob
-            const copy = view.slice() as Uint8Array;
-            controller.enqueue(copy);
-        }
-        pull_promise_control.resolve();
-    }
-    catch (err) {
-        pull_state.dispose();
-        pull_promise_control.reject(err);
-    }
-    finally {
-        pull_state.__pull_promise_control = null;
-    }
+export function http_wasm_create_transform_stream(): TransformStreamExtension {
+    const transform_stream = new TransformStream<Uint8Array, Uint8Array>() as TransformStreamExtension;
+    transform_stream.__writer = transform_stream.writable.getWriter();
+    return transform_stream;
 }
 
-export function http_wasm_readable_stream_controller_error(pull_state: PullStateExtension, error: Error): void {
-    const controller = pull_state.__controller;
-    mono_assert(controller, "expected controller");
-    pull_state.__pull_promise_control?.reject(error);
-    pull_state.__fetch_promise_control?.reject(error);
-    pull_state.dispose();
-    controller.error(error);
-    pull_state.__pull_promise_control = null;
+export async function http_wasm_transform_stream_write(ts: TransformStreamExtension, bufferPtr: VoidPtr, bufferLength: number): Promise<void> {
+    mono_assert(bufferLength > 0, "expected bufferLength > 0");
+    mono_assert(ts.__fetch_promise_control, "expected fetch promise control");
+    await Promise.race([ts.__writer.ready, ts.__fetch_promise_control.promise]);
```

Comment:
Race with fetch everywhere to get errors and cancellation because TransformStream does not propagate these.

Reply:
If fetch failed/aborted/whatever it would throw and we could catch it here ? And abort the other promise via it's controller too ?

Reply:
transformer ends have `cancel()` and `abort()`

Reply:
We could still wrap those promises with cancelable and have our own controller

Reply:
`ts.__writer.abort(error);` is not locked after all ?

Anyway, that solves the case when we are aborting it. I think the case when TCP fails is still hanging ?

Reply:
> ts.__writer.abort(error); is not locked after all ?

Everything on the writer is not locked because we created it.
Only the `ReadableStream` is locked because the browser is using it.

I tried forwarding fetch errors to `writer.abort(error)` but it doesn't work in all cases. E.g.

```js
var t = new TransformStream();
var w = t.writable.getWriter();
var a = new AbortController();
var f = fetch("", {
    method: "POST",
    body: t.readable,
    duplex: "half",
    signal: a.signal
});
f.catch(error => w.abort(error));

var wp = w.ready.then(() => { a.abort(); return w.write(new Uint8Array([1, 2, 3])); });

// wp is pending and never resolves
// f will be rejected
```

The `BrowserHttpHandler_StreamingRequest_Http1Fails` would fail because the `write()` promise never resolves as fetch gets rejected during that write...